cpp/src/merge/merge.cu

```diff
@@ -0,0 +1,489 @@
+ * according to `merged_indices` map such that bit `i` in `out_col`
+ * will be equal to bit `thrust::get<1>(merged_indices[i])` from `left_dcol`
+ * if `thrust::get<0>(merged_indices[i])` equals `side::LEFT`; otherwise,
+ * from `right_dcol`.
+ *
+ * `left_dcol`, `right_dcol` and `out_dcol` must not
+ * overlap.
+ *
+ * @tparam left_have_valids Indicates whether left_dcol mask is unallocated (hence, ALL_VALID)
+ * @tparam right_have_valids Indicates whether right_dcol mask is unallocated (hence ALL_VALID)
+ * @param[in] left_dcol The left column_device_view whose bits will be merged
+ * @param[in] right_dcol The right column_device_view whose bits will be merged
+ * @param[out] out_dcol The output mutable_column_device_view after merging the left and right
+ * @param[in] num_destination_rows The number of rows in the out_dcol
+ * @param[in] merged_indices The map that indicates the source of the input and index
+ * to be copied to the output. Length must be equal to `num_destination_rows`
+ */
+template <bool left_have_valids, bool right_have_valids>
+__global__ void materialize_merged_bitmask_kernel(cudf::column_device_view left_dcol,
+                                                  cudf::column_device_view right_dcol,
+                                                  cudf::mutable_column_device_view out_dcol,
+                                                  cudf::size_type const num_destination_rows,
+                                                  index_type const* const __restrict__ merged_indices) {
+  cudf::size_type destination_row = threadIdx.x + blockIdx.x * blockDim.x;
+
+  cudf::bitmask_type const* const __restrict__ source_left_mask = left_dcol.null_mask();
+  cudf::bitmask_type const* const __restrict__ source_right_mask= right_dcol.null_mask();
+  cudf::bitmask_type* const __restrict__ destination_mask = out_dcol.null_mask();
+  
+  auto active_threads =
+    __ballot_sync(0xffffffff, destination_row < num_destination_rows);
+
+  while (destination_row < num_destination_rows) {
+    index_type const& merged_idx = merged_indices[destination_row];
+    side const src_side = thrust::get<0>(merged_idx);
+    cudf::size_type const src_row  = thrust::get<1>(merged_idx);
+    bool const from_left{src_side == side::LEFT};
+    bool source_bit_is_valid{true};
+    if (left_have_valids && from_left) {
+      source_bit_is_valid = left_dcol.is_valid_nocheck(src_row);
+    }
+    else if (right_have_valids && !from_left) {
+      source_bit_is_valid = right_dcol.is_valid_nocheck(src_row);
+    }
+
+    // Use ballot to find all valid bits in this warp and create the output
+    // bitmask element
+    cudf::bitmask_type const result_mask{
+      __ballot_sync(active_threads, source_bit_is_valid)};
+
+    cudf::size_type const output_element = cudf::word_index(destination_row);
+
+    // Only one thread writes output
+    if (0 == threadIdx.x % warpSize) {
+      destination_mask[output_element] = result_mask;
```

Comment:
I think you could use `out_dcol.set_mask_word()` here

Reply:
Done.